main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1496,13 +1516,99 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
   }
 
   /**
-    * Tries to apply `tpe1` to `tpe2`. Creates a resolution error if `tpe1` is not a type constructor.
+    * Create a well-formed type applying `tpe1` to `tpe2`.
     */
   private def mkApply(tpe1: Type, tpe2: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
-    tpe1.kind match {
-      case _: Kind.Arrow => Type.Apply(tpe1, tpe2).toSuccess
+    (tpe1.kind, tpe2.kind)  match {
+      case (Kind.Arrow(k1, _), k2) if k2 <:: k1 => Type.Apply(tpe1, tpe2).toSuccess
       case _ => ResolutionError.IllegalTypeApplication(tpe1, tpe2, loc).toFailure
     }
   }
 
+  /**
+    * Create a well-formed type applying `tpe` to `args`.
+    */
+  private def mkApply(tpe: Type, args: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    fold(args, tpe)(mkApply(_, _, loc))
```

Comment:
fold -> Validation.fold (that way it is easier to see what kind of fold this is imho).